src/main/java/com/aws/greengrass/clientdevices/auth/certificate/CISShadowMonitor.java

```diff
@@ -217,56 +225,74 @@ private void processCISShadow(ShadowDeltaUpdatedEvent event) {
         // NOTE: This method executes in an MQTT CRT thread. Since certificate generation is a compute intensive
-        // operation (particularly on low end devices) it is imperative that we process this event asynchronously
+        // operation (particularly on low-end devices) it is imperative that we process this event asynchronously
         // to avoid blocking other MQTT subscribers in the Nucleus
         CompletableFuture.runAsync(() -> {
-            try {
-                RetryUtils.runWithRetry(GET_CONNECTIVITY_RETRY_CONFIG, connectivityInfoProvider::getConnectivityInfo,
-                        "get-connectivity", LOGGER);
-            } catch (InterruptedException e) {
-                LOGGER.atDebug().kv(VERSION, version).cause(e)
-                        .log("Retry workflow for getting connectivity info interrupted");
-                Thread.currentThread().interrupt();
-                return;
-            } catch (Exception e) {
-                LOGGER.atError().kv(VERSION, version).cause(e)
-                        .log("Failed to get connectivity info from cloud."
-                                + " Check that the core device's IoT policy "
-                                + "grants the greengrass:GetConnectivityInfo permission");
-                return;
-            }
+            synchronized (shadowProcessingLock) {
+                LOGGER.atInfo().log("Processing CIS version: {}", version);
 
-            try {
-                for (CertificateGenerator cg : monitoredCertificateGenerators) {
-                    cg.generateCertificate(connectivityInfoProvider::getCachedHostAddresses,
-                            "connectivity info was updated");
+                List<String> newHostAddresses;
+                try {
+                    newHostAddresses = fetchHostAddressesFromCIS();
+                } catch (InterruptedException e) {
+                    LOGGER.atDebug().kv(VERSION, version).cause(e)
+                            .log("Retry workflow for getting connectivity info interrupted");
+                    Thread.currentThread().interrupt();
+                    return;
+                } catch (Exception e) {
+                    LOGGER.atError().kv(VERSION, version).cause(e)
+                            .log("Failed to get connectivity info from cloud."
+                                    + " Check that the core device's IoT policy "
+                                    + "grants the greengrass:GetConnectivityInfo permission");
+                    return;
                 }
-            } catch (CertificateGenerationException e) {
-                LOGGER.atError().kv(VERSION, version).cause(e).log("Failed to generate new certificates");
-                return;
-            }
 
-            try {
-                updateCISShadowReportedState(version, desiredState);
-            } finally {
-                lastVersion = version;
+                List<String> previousHostAddresses = hostAddresses;
```

Comment:
Should we use a set instead? We don't want duplicates and we don't really care about the order.